Add WC Tracker unique payment method

menakas opened the pull request:
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

The WC Tracker client code groups orders(counts and total volume) by gateway(payment method) and also by origin.
But at times, these gateway(payment method) names or the origin names contain some unique id like `Visa **** 1635` defeating the purpose of the grouping. This PR attempts to remove such unique ids formed using digits as a first step in consolidating the order details.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create some orders that have the `_payment_method` meta set to a value with a unique id. If this cannot be done by placing an order, then change the value in existing orders in the database.
<img width="662" alt="Screenshot 2023-04-24 at 4 46 16 PM" src="https://user-images.githubusercontent.com/6427897/234000265-637d018d-1ba1-4677-ab53-55a90c0b9ec2.png">

2. Trigger a Tracker snapshot - Call `send_tracking_data()`. Check the values in the snapshot created. Alternately, call `get_tracking_data()` and look at the results.


3. On trunk, the same payment method with different reference ids will be listed as different items in the meta. On this branch, they should be combined into one with the unique ids stripped off.
<img width="512" alt="Screenshot 2023-04-24 at 4 48 14 PM" src="https://user-images.githubusercontent.com/6427897/234000788-9985dffc-42e8-4027-941c-c4de3dc2cae8.png">

4. Do a similar process for `created_via`. 


<img width="249" alt="Screenshot 2023-04-24 at 4 52 36 PM" src="https://user-images.githubusercontent.com/6427897/234001732-c86a3e50-a2dd-454f-b448-a758846c3672.png">




<!-- End testing instructions -->

menakas committed:
create a key of gateway name in lowercase. Remove standard keywords in it

menakas committed:
if the key contains a digit, truncate upto that point, especially if the count is very less

menakas committed:
now group counts and totals by the new key

menakas committed:
Fix some Code sniff errors

deltaWhiskey commented on plugins/woocommerce/includes/class-wc-tracker.php:
```diff
@@ -568,7 +584,21 @@ private static function get_orders_origins() {
+			$key = strtolower( $origin->origin );
+
+			// If the origin name has a unique id, discard it, especially if the count is 1.
+			if ( 1 === $origin->count ) {
```
```suggestion
			if ( 1 == $origin->count ) {
```
In my test environment, `$origin->count` is a string, so it never `===` 1.

menakas replied:
Thanks Dan. I've added a commit to use loose comparison.

deltaWhiskey reviewed (dismissed):
I like this approach. It errs on the side of collecting less detail, and that's best for everyone.

An edge case: payment methods like `przelewy24` will be `przelewy24` when there are lots of orders, but `przelewy` when there are just a few. I think that's OK, though.

This code does a good job of avoiding replacing anything when the first character is a digit,  like `2checkout`.

I have a theoretical concern about methods like `x2pay` or `transaction 123456 via Some Method`. I just made those examples up, though.

menakas committed:
Compare after casting to int

menakas commented:
>I have a theoretical concern about methods like x2pay or transaction 123456 via Some Method. I just made those examples up, though.

@deltaWhiskey Same here. I did think of payment methods that might have such names. 
Short of writing an intelligent unique id detector, I could not think of a cover-all solution.

One additional condition could be that the first digit's position should be at least after some arbitrary n characters so that we definitely retain the first n characters, hoping that something like `transaction 123456` does not occur.

Or just strip off all digits. That would mean that unique ids with alphanumeric chars will still produce unique payment methods. eg: Created Via PSP Api(227543ID1679611759) - This is a real one, not made up.

menakas commented:
@deltaWhiskey I have been working on other options for doing this grouping more effectively. I think I have an idea. I'll try to work it out tomorrow.

menakas committed:
Add a function that groups a array of object by removing unique ids

menakas committed:
Convert array into an associative array

menakas committed:
Convert array into an associative array for gateways as well

menakas committed:
Manipulate key to remove currecy as a prefix

menakas committed:
Manipulate key for orders by origin as well

menakas commented:
@deltaWhiskey 
ok, here is the general logic. 
1. Convert the array of objects (with properties like `gateway`, `count`, `total` and `currency`) into an associative array with the `gateway` as the key.
2. Take this array's keys and sort them by length. Within the same length of keys, sort by name. This would mean that similar payment methods with just the unique ids different are adjacent.
4. Consider each adjacent pair of keys. Tokenize them with standard delimiters ( space, comma, hyphen, underscore, colon at this point).
5. Look for the count of tokens that are the same. If only one token is different and not the first token, then they should be grouped by a key that doesn't contain that token. So, add a group key from this set of common tokens to this object's `group_key` property.
6. Now use this new property of the objects to group the counts and totals.

Same for `orders by origin`.

I think this would cast a wider net. One case that can fail is if the unique id is the first token of the payment method. Or if the unique id is more than a token long, say it has a hyphen in-between.

Let me know if this can be further refined.

I've tested this locally on a large set of data as a CSV file. Also on a small local store. However, I'd like to verify it on a large test store tomorrow.

menakas committed:
Fix Code Sniffer errors

deltaWhiskey commented:
The new logic is a big improvement.

> If only one token is different and not the first token, then they should be grouped by a key that doesn't contain that token.

If I'm reading the code right, the different token will be omitted even if it's the first token. So  `#12345 SomeGateway` becomes `SomeGateway`. That's a good thing.

You bring up a good point about unique IDs with hyphens, but this still looks like a good solution.

deltaWhiskey commented on plugins/woocommerce/includes/class-wc-tracker.php:
```diff
@@ -530,14 +588,47 @@ private static function get_orders_by_gateway() {
 
 		$orders_by_gateway_currency = array();
+
+		$orders_by_gateway = self::group_objects_by_key(
```
I think we should give some context why this logic exists:

```suggestion
		// Group data by payment method. Strip any transaction identifiers from method descriptions.
		$orders_by_gateway = self::group_objects_by_key(
```

menakas replied:
Sure will give a detailed explanation.

menakas commented:
I've tested it with sample large data on the local store. I'll need to test the orders by origin as well.

menakas committed:
Add Changelog entry

menakas committed:
Refine comments for the logic

menakas commented:
@deltaWhiskey 
I added the changelog entry 🎆 
I've added somewhat more meaningful comments explaining the logic.
I've tested both the payment methods and origin with a large partially real dataset. More info here p2-p7Ldg5-44W

There is still scope for some pre/post-processing to get better accuracy, but I'm wary of overdoing it as real data could be really weird.

Let me know your thoughts!

deltaWhiskey commented on plugins/woocommerce/includes/class-wc-tracker.php:
```diff
@@ -498,6 +498,66 @@ private static function get_order_dates() {
+				$len_prev = count( $prev_tokens );
+
+				// Gather the common tokens.
+				// Let us allow for the unique reference id to be anywhere in the name but for the first token.
```
Is the "but for the first token" part of this comment accurate? I don't see where the first token is handled differently.

menakas replied:
That is a remnant of a logic I wrote earlier. I've removed it.

menakas committed:
Add logic to check if the differing token contains a digit

menakas commented:
Made another change to look for a digit in the tokens that differ.

menakas commented:
@coreymckrill would be mind providing your feedback from the POV of proton?

coreymckrill commented on plugins/woocommerce/includes/class-wc-tracker.php:
```diff
@@ -498,6 +498,69 @@ private static function get_order_dates() {
+				$comm_tokens = array();
+
+				// Tokenize the current and previous gateway names.
+				$curr_tokens = preg_split( '/[ :,\-_]/ ', $key );
+				$prev_tokens = preg_split( '/[ :,\-_]/ ', $prev );
```
```suggestion
				$curr_tokens = preg_split( '/[ :,\-_]+/', $key );
				$prev_tokens = preg_split( '/[ :,\-_]+/', $prev );
```

Might be worth detecting more than one token boundary at a time. For example, if we have something like `PayPalTransaction ID: #12345`, the `:` and the ` ` could be considered one boundary instead of two separate ones.

My test data looks like this:

```
cheque: *123
cheque: #456
cheque  #8675
```

Which results in:

```
  ["gateway_cheque #8675_USD_count"]=>
  string(1) "1"
  ["gateway_cheque #8675_USD_total"]=>
  string(6) "325.49"
  ["gateway_cheque _USD_count"]=>
  int(2)
  ["gateway_cheque _USD_total"]=>
  float(650.98)
```

But with the change above, I get:

```
  ["gateway_cheque_USD_count"]=>
  int(3)
  ["gateway_cheque_USD_total"]=>
  float(976.47)
```

Could also take this further here or later on and `trim` the tokens to ensure that resulting keys are more normalized.

menakas replied:
Thanks @coreymckrill 
That sounds like a good improvisation. I'm committing the suggestion.

coreymckrill reviewed (dismissed):
👍 Looks good to me. Works as described in the testing instructions (I edited the database directly and used completely made up strings for the payment methods).

One suggestion below that might help refine the group key extraction further.

Also just noting that the WC Tracker stuff does not seem to be HPOS compatible. Nothing to do about that here in this PR, but I opened a [separate issue](https://github.com/woocommerce/woocommerce/issues/38148) for it.

menakas committed:
Squash multiple delimiters of tokens into one

Co-authored-by: Corey McKrill <916023+coreymckrill@users.noreply.github.com>

menakas committed:
Further trim the special characters from the key

menakas commented:
I've also added another commit to trim the special characters and space from the final key.

@deltaWhiskey let me know if this looks good.

deltaWhiskey reviewed (approved):
Looks good!

menakas commented:
@coreymckrill I've incorporated the changes you suggested. 
Would you mind merging this PR if it looks fine? I do not have the authorization to merge it.
Thanks.
